c++/src/H5DataSet.cpp

```diff
@@ -76,6 +76,27 @@ DataSet::DataSet(const DataSet &original) : H5Object(), AbstractDs(), id(origina
     incRefCount(); // increment number of references to this id
 }
 
+//--------------------------------------------------------------------------
+// Function:    DataSet assignment operator
+///\brief       Assignment operator: same HDF5 object as \a original
+///\param       original - IN: DataSet instance to copy
+// Programmer   Lee Newberg - 2021
+//--------------------------------------------------------------------------
+DataSet &
+DataSet::operator=(const DataSet &original)
+{
+    // Wrap the changes to `*this` by an incremented reference count for
+    // `original` to prevent trouble in the case that `*this` and
+    // `original` share reference counts or memory in any relevant
+    // sense.
+    original.incRefCount();
+    decRefCount(); // for old value of id
+    id = original.id;
+    incRefCount(); // for new value of id
+    original.decRefCount();
+    return *this;
+}
+
 //--------------------------------------------------------------------------
```

Comment:
Please consider using IdComponent::setId() to ensure the current dataset ID is properly closed.  setId() also handles the ref counting.  Perhaps, DataType::operator= can be used as an example.  Thank you!

Reply:
@bmribler Thank you for the feedback.  The new code has `H5DataSet::operator=` mimicking `H5DataType::operator=` as you suggest.

Reply:
If someone could contribute a test that would be great.  I am a little over my head there.

Reply:
Thank you, @Leengit.  I'll try to come up with a test.